Emit dropck normalization errors in borrowck

matthewjasper opened the pull request:
Borrowck generally assumes that any queries it runs for type checking will succeed, thinking that HIR typeck will have errored first if there was a problem. However as of #98641, dropck isn't run on HIR, so there's no direct guarantee that it doesn't error. While a type being well-formed might be expected to ensure that its fields are well-formed, this is not the case for types containing a type projection:

```rust
pub trait AuthUser {
    type Id;
}

pub trait AuthnBackend {
    type User: AuthUser;
}

pub struct AuthSession<Backend: AuthnBackend> {
    data: Option<<<Backend as AuthnBackend>::User as AuthUser>::Id>,
}

pub trait Authz: Sized {
    type AuthnBackend: AuthnBackend<User = Self>;
}

pub fn run_query<User: Authz>(auth: AuthSession<User::AuthnBackend>) {}
// ^ No User: AuthUser bound is required or inferred.
```

While improvements to trait solving might fix this in the future, for now we go for a pragmatic solution of emitting an error from borrowck (by rerunning dropck outside of a query) and making drop elaboration check if an error has been emitted previously before panicking for a failed normalization.

Closes #103899
Closes #135039

r? @compiler-errors (feel free to re-assign)

matthewjasper committed:
Report dropck normalization errors in borrowck

HIR type checking no longer runs dropck, so we may get new errors when
we run it in borrowck. If this happens then rerun the query in a local
infcx and report errors for it.

matthewjasper committed:
Handle normalization failures in drop elaboration

Drop elaboration looks at fields of a type, which may error when we try
to normalize them. Borrowck will have detected this if HIR typeck
didn't, but we don't delete the MIR body for errors in borrowck so
still have to handle this happening in drop elaboration by checking
whether an error has been emitted.

matthewjasper committed:
Update tests for dropck normalization errors

Takes crash tests from #135039, #103899, #91985 and #105299 and turns them into ui tests

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13136849859/job/36654170511) [(plain)](https://github.com/rust-lang/rust/commit/abfa754e4470f12359f9b5ea97f121a791b2424b/checks/36654170511/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#22 exporting to docker image format
#22 sending tarball 28.1s done
#22 DONE 34.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
##[endgroup]
##[group]Testing stage2 compiletest suite=crashes mode=crashes (x86_64-unknown-linux-gnu)

running 230 tests
2025-02-04T14:07:39.289232Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge. If you want to see verbose output, set `COMPILETEST_VERBOSE_CRASHES=1`."
   [crashes] tests/crashes/103899.rs ... FAILED
2025-02-04T14:07:39.296121Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge. If you want to see verbose output, set `COMPILETEST_VERBOSE_CRASHES=1`."
   [crashes] tests/crashes/105299.rs ... FAILED
20%  --  47/230,  45 passed, 2 failed, 0 ignored
30%  --  70/230,  68 passed, 2 failed, 0 ignored
40%  --  93/230,  91 passed, 2 failed, 0 ignored
50%  -- 116/230, 114 passed, 2 failed, 0 ignored
50%  -- 116/230, 114 passed, 2 failed, 0 ignored
60%  -- 139/230, 137 passed, 2 failed, 0 ignored
70%  -- 162/230, 160 passed, 2 failed, 0 ignored
80%  -- 184/230, 182 passed, 2 failed, 0 ignored
2025-02-04T14:07:41.527648Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge. If you want to see verbose output, set `COMPILETEST_VERBOSE_CRASHES=1`."
   [crashes] tests/crashes/135039.rs ... FAILED
90%  -- 207/230, 204 passed, 3 failed, 0 ignored
2025-02-04T14:07:41.641403Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge. If you want to see verbose output, set `COMPILETEST_VERBOSE_CRASHES=1`."
   [crashes] tests/crashes/91985.rs ... FAILED


failures:


---- [crashes] tests/crashes/103899.rs stdout ----

error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge. If you want to see verbose output, set `COMPILETEST_VERBOSE_CRASHES=1`.
thread '[crashes] tests/crashes/103899.rs' panicked at src/tools/compiletest/src/runtest/crashes.rs:17:18:
fatal error
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


---- [crashes] tests/crashes/105299.rs stdout ----

error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge. If you want to see verbose output, set `COMPILETEST_VERBOSE_CRASHES=1`.
thread '[crashes] tests/crashes/105299.rs' panicked at src/tools/compiletest/src/runtest/crashes.rs:17:18:
fatal error

---- [crashes] tests/crashes/135039.rs stdout ----
---- [crashes] tests/crashes/135039.rs stdout ----

error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge. If you want to see verbose output, set `COMPILETEST_VERBOSE_CRASHES=1`.
thread '[crashes] tests/crashes/135039.rs' panicked at src/tools/compiletest/src/runtest/crashes.rs:17:18:
fatal error

---- [crashes] tests/crashes/91985.rs stdout ----
---- [crashes] tests/crashes/91985.rs stdout ----

error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge. If you want to see verbose output, set `COMPILETEST_VERBOSE_CRASHES=1`.
thread '[crashes] tests/crashes/91985.rs' panicked at src/tools/compiletest/src/runtest/crashes.rs:17:18:
fatal error


```

</details>

compiler-errors commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit abfa754e4470f12359f9b5ea97f121a791b2424b with merge 40f1f75bb4cd41066f6e5b10d25251675eacd0d7...
<!-- homu: {"type":"TryBuildStarted","head_sha":"abfa754e4470f12359f9b5ea97f121a791b2424b","merge_sha":"40f1f75bb4cd41066f6e5b10d25251675eacd0d7"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13138347134/job/36666476720)
Build commit: 40f1f75bb4cd41066f6e5b10d25251675eacd0d7 (`40f1f75bb4cd41066f6e5b10d25251675eacd0d7`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13138347134/job/36666476720"},"merge_sha":"40f1f75bb4cd41066f6e5b10d25251675eacd0d7"} -->

rust-timer commented:
Queued 40f1f75bb4cd41066f6e5b10d25251675eacd0d7 with parent 7b31983d5506e975801963b88b36c504eb2cecd8, future [comparison URL](https://perf.rust-lang.org/compare.html?start=7b31983d5506e975801963b88b36c504eb2cecd8&end=40f1f75bb4cd41066f6e5b10d25251675eacd0d7).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (40f1f75bb4cd41066f6e5b10d25251675eacd0d7): [comparison URL](https://perf.rust-lang.org/compare.html?start=7b31983d5506e975801963b88b36c504eb2cecd8&end=40f1f75bb4cd41066f6e5b10d25251675eacd0d7&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=7b31983d5506e975801963b88b36c504eb2cecd8&end=40f1f75bb4cd41066f6e5b10d25251675eacd0d7&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=7b31983d5506e975801963b88b36c504eb2cecd8&end=40f1f75bb4cd41066f6e5b10d25251675eacd0d7&stat=max-rss)
<details>
<summary>Results (secondary -0.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.7% | [-0.7%, -0.7%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=7b31983d5506e975801963b88b36c504eb2cecd8&end=40f1f75bb4cd41066f6e5b10d25251675eacd0d7&stat=cycles:u)
<details>
<summary>Results (secondary 3.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.3% | [3.3%, 3.3%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=7b31983d5506e975801963b88b36c504eb2cecd8&end=40f1f75bb4cd41066f6e5b10d25251675eacd0d7&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 778.8s -> 781.102s (0.30%)
**Artifact size**: 328.88 MiB -> 328.83 MiB (-0.02%)

compiler-errors commented:
@craterbot check

matthewjasper commented:
@craterbot run start=master end=try#40f1f75bb4cd41066f6e5b10d25251675eacd0d7

lqd commented:
@matthewjasper that's build-and-test

matthewjasper commented:
@craterbot abort

matthewjasper commented:
@craterbot run start=master end=try#40f1f75bb4cd41066f6e5b10d25251675eacd0d7 mode=check-only

compiler-errors commented:
@craterbot cancel

compiler-errors commented:
@bors try

bors commented:
:hourglass: Trying commit defefbc030e4b2d1ddbcaa07cbbc49d26fd66a48 with merge 35b6491ce6d2b388d0a5de1e59e2123604af807f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"defefbc030e4b2d1ddbcaa07cbbc49d26fd66a48","merge_sha":"35b6491ce6d2b388d0a5de1e59e2123604af807f"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13210522184/job/36885422987)
Build commit: 35b6491ce6d2b388d0a5de1e59e2123604af807f (`35b6491ce6d2b388d0a5de1e59e2123604af807f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13210522184/job/36885422987"},"merge_sha":"35b6491ce6d2b388d0a5de1e59e2123604af807f"} -->

compiler-errors commented:
@craterbot check

compiler-errors commented on compiler/rustc_trait_selection/src/traits/query/dropck_outlives.rs:
```diff
@@ -146,14 +162,17 @@ pub fn compute_dropck_outlives_inner<'tcx>(
             &mut constraints,
-        )?;
+        ) {
+            Err(_) => return Err(Vec::new()),
```
Do we ever actually hit this path? I wonder if we should just delay a bug `dtorck_constraint_for_ty_inner` and return `Ok` -- seems to only return `NoSolution` for placeholder/infer/error.

compiler-errors commented on compiler/rustc_trait_selection/src/traits/query/dropck_outlives.rs:
```diff
@@ -173,11 +192,25 @@ pub fn compute_dropck_outlives_inner<'tcx>(
+                    return Err(errors);
+                }
+                normalized_ty
+            } else if let Ok(Normalized { value: ty, obligations }) =
```
Could you invert this? First use query normalization, otherwise `if report_errors` recompute the errors from deeply normalizing, or otherwise return an empty vec?

Also, does it really matter if we recompute the errors unconditionally? Presumably we're never computing dropck outlives and expecting it to fail even if the program passes, right?

compiler-errors commented on compiler/rustc_borrowck/src/type_check/liveness/trace.rs:
```diff
@@ -589,19 +594,45 @@ impl<'tcx> LivenessContext<'_, '_, '_, 'tcx> {
+                if !errors.is_empty() {
+                    typeck.infcx.err_ctxt().report_fulfillment_errors(errors);
+                } else {
+                    rustc_middle::span_bug!(span, "Rerunning drop data query produced no error.");
```
import span_bug

matthewjasper committed:
Clean up dropck code a bit

- Remove `Result` that couldn't be Err on valid compilation.
- Always compute errors on failure.

bors commented:
:umbrella: The latest upstream changes (presumably #137030) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

compiler-errors commented:
whoops, r=me after rebase

bors commented:
:pushpin: Commit 49cf00c7c04e24bb27a09fbcfe9bac9278075608 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 49cf00c7c04e24bb27a09fbcfe9bac9278075608 -->
<!-- homu: {"type":"Approved","sha":"49cf00c7c04e24bb27a09fbcfe9bac9278075608","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 49cf00c7c04e24bb27a09fbcfe9bac9278075608 with merge ed49386d3aa3a445a9889707fd405df01723eced...
<!-- homu: {"type":"BuildStarted","head_sha":"49cf00c7c04e24bb27a09fbcfe9bac9278075608","merge_sha":"ed49386d3aa3a445a9889707fd405df01723eced"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13407753668/job/37460340628)
Approved by: compiler-errors
Pushing ed49386d3aa3a445a9889707fd405df01723eced to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13407753668/job/37460340628"},"merge_sha":"ed49386d3aa3a445a9889707fd405df01723eced"} -->

rust-timer commented:
Finished benchmarking commit (ed49386d3aa3a445a9889707fd405df01723eced): [comparison URL](https://perf.rust-lang.org/compare.html?start=5986ff05d8480da038dd161b3a6aa79ff364a851&end=ed49386d3aa3a445a9889707fd405df01723eced&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=5986ff05d8480da038dd161b3a6aa79ff364a851&end=ed49386d3aa3a445a9889707fd405df01723eced&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=5986ff05d8480da038dd161b3a6aa79ff364a851&end=ed49386d3aa3a445a9889707fd405df01723eced&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=5986ff05d8480da038dd161b3a6aa79ff364a851&end=ed49386d3aa3a445a9889707fd405df01723eced&stat=cycles:u)
<details>
<summary>Results (primary 4.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.3% | [1.7%, 6.4%] | 7     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 4.3% | [1.7%, 6.4%] | 7     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=5986ff05d8480da038dd161b3a6aa79ff364a851&end=ed49386d3aa3a445a9889707fd405df01723eced&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 773.598s -> 773.063s (-0.07%)
**Artifact size**: 360.32 MiB -> 362.34 MiB (0.56%)
